hdk-proc-macros/src/to_tokens.rs

```diff
@@ -0,0 +1,141 @@
+                    Err(allocation_error) => return hdk::holochain_core_types::error::RibosomeEncodedValue::from(allocation_error).into(),
+                };
+                let init = hdk::global_fns::init_global_memory(allocation);
+                if init.is_err() {
+                    return hdk::holochain_wasm_utils::memory::ribosome::return_code_for_allocation_result(
+                        init
+                    ).into();
+                }
+
+                // Macro'd InputStruct
+                #[derive(Deserialize, Serialize, Debug, hdk::holochain_core_types_derive::DefaultJson)]
+                struct InputStruct {
+                    #(#input_params),*
+                }
+
+                // Deserialize input
+                let input: InputStruct = hdk::load_json!(encoded_allocation_of_input);
+
+                // Macro'd function body
+                fn execute (params: InputStruct) #output_param_type {
+                    let InputStruct { #(#input_param_names),* } = params;
+                    #function_body
+                }
+
+                hdk::holochain_wasm_utils::memory::ribosome::return_code_for_allocation_result(
+                    hdk::global_fns::write_json(execute(input))
+                ).into()
+            }
+        })
+    }
+}
+
+impl ToTokens for FnParameter {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let input_param_name = &self.ident;
+
+        let input_param_type = &self.ty;
+
+        tokens.extend(quote! {
+            FnParameter::new(stringify!(#input_param_name), stringify!(#input_param_type))
+        })
+    }
+}
+
+impl ToTokens for FnDeclaration {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let zome_function_name = &self.name;
+        let input_params = &self.inputs;
+        let output_params = match &self.output {
+            syn::ReturnType::Default => Vec::new(),
+            syn::ReturnType::Type(_, ty) => {
+                vec![quote!(FnParameter::new("result", stringify!(#ty)))]
+            }
+        };
+
+        tokens.extend(quote! {
+            FnDeclaration {
+                name: #zome_function_name.to_string(),
+                inputs: vec![#(#input_params,)*],
+                outputs: vec![#(#output_params,)*],
+            }
+        })
+    }
+}
+
+impl ToTokens for ZomeCodeDef {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let genesis = self.genesis();
+        let zome_setup = self.zome_setup();
+        let list_traits = self.list_traits();
+        let list_functions = self.list_functions();
+        let zome_fns = self.zome_fns.clone();
+        let entry_def_fns = self.entry_def_fns.clone();
+        let extra = &self.extra;
+        let receive_callback = self.receive_callback();
+        let panic_handler = panic_handler();
+
+        tokens.extend(quote! {
+
+            #(#extra)*
+
+            #(#entry_def_fns )*
+
+            #genesis
```

Comment:
small thing: when this gets merged it will need to be updated to `init` and I'd suggest we make it be `init_callback` to match `receive_callback.`